app/ai/voice/agents/breeze_buddy/managers/data_source_prefetch.py

```
@@ -0,0 +1,167 @@
+_CACHE_TTL = 60  # seconds — shared across leads; short to keep data fresh
+_FETCH_TIMEOUT = 5.0
+
+
+def _cache_key(ref: DataSourceRef) -> str:
+    """Derive a stable cache key from the sheet config."""
+    spreadsheet_id = extract_spreadsheet_id(ref.spreadsheet_url)
+    if not spreadsheet_id:
+        spreadsheet_id = "invalid_url"
+    sheet = ref.sheet_name or "_first_"
+    cols = json.dumps(sorted(ref.columns or []))
+    cols_digest = hashlib.md5(cols.encode()).hexdigest()[:8]
```

Ruff (0.15.17)

[error] Probable use of insecure hash functions in `hashlib`: `md5`

(S324)